src/pykeen/nn/emb.py

```diff
@@ -1068,3 +1070,103 @@ def forward(
         if indices is not None:
             x = x[indices]
         return x
+
+
+class NodePieceRepresentation(RepresentationModule):
+    """
+    Basic implementation of node piece decomposition [galkin2021]_.
+
+    .. note ::
+        This implementation currently only supports representation of entities by bag-of-relations.
+    """
+
+    #: the token representations
+    tokens: torch.FloatTensor
+
+    #: the entity-to-token mapping
+    assignment: torch.LongTensor
+
+    def __init__(
+        self,
+        *,
+        triples_factory: CoreTriplesFactory,
+        token_representation: EmbeddingSpecification,
+        aggregation: Callable[[torch.Tensor, int], torch.Tensor] = None,
+        k: int = 1,
+        **kwargs,
+    ):
+        """
+        Initialize the representation.
+
+        :param triples_factory:
+            the triples factory
+        :param token_representation:
+            the token representation specification
+        :param aggregation:
+            aggregation of multiple token representations to a single entity representation
+        :param k:
+            the number of tokens for each entity.
+        """
+        if triples_factory.create_inverse_triples:
+            raise NotImplementedError("Triples factories with inverse triples are currently not supported.")
+
+        # create token representations
+        tokens = token_representation.make(
+            num_embeddings=2 * triples_factory.num_relations + 1,  # normal relations + inverse relations + padding
+        )
+
+        # super init; has to happen *before* any parameter or buffer is assigned
+        super().__init__(max_id=triples_factory.num_entities, shape=tokens.shape, **kwargs)
+
+        # normalize aggregation
+        if aggregation is None:
+            aggregation = torch.mean
+        self.aggregation = aggregation
+
+        # assign module
+        self.tokens = tokens
+
+        # tokenize: represent entities by bag of relations
+        h, r, t = triples_factory.mapped_triples.t()
+
+        # collect candidates
+        e2r = defaultdict(set)
+        for e, r in (
+            torch.cat(
+                [
+                    torch.stack([h, r], dim=1),
+                    torch.stack([t, r + triples_factory.num_relations], dim=1),
+                ],
+                dim=0,
+            )
+            .unique(dim=0)
+            .tolist()
+        ):
+            e2r[e].add(r)
+
+        # randomly sample with replacement k relations for each entity
+        assignment = torch.full(
+            size=(triples_factory.num_entities, k),
+            dtype=torch.long,
+            fill_value=2 * triples_factory.num_relations,  # padding ID
+        )
+        for e, rs in e2r.items():
+            assignment[e] = torch.as_tensor(data=random.choices(population=list(rs), k=k), dtype=torch.long)
```

Comment:
can torch not handle random sampling? this introduces yet another random thing we'd have to control for reproducibility :/

Reply:
```suggestion
            rs = torch.as_tensor(data=list(rs), dtype=torch.long)
            assignment[e] = rs[torch.randint(high=rs.numel(), size=(k,))]
```

Reply:
Done in f771e7ae